### PR TITLE
SignOn Messages

### PR DESCRIPTION
**Pull request by zachferland**

### CODE REVIEW

File: src/chat/msg.rs [hunk 1]
```diff
@@ -20,10 +23,20 @@ pub type Result<T> = result::Result<T, Error>;
...
-    Signon(String, String, PublicKey, SocketAddrV4),
+    /// Time, username, SHA-2(username | password), PKc1, ip. Time must be
+    /// fairly recent or the server will reject
+    //TODO:  Is a timestamp necessary? replays don't necessarily have any
+    //       harmful effects. It would only make someone apear as if they
+    //       are online when they really are not.
+    //TODO:  Is there any vulnerability in sending the same auth challenge
+    //       everytime?
+    //NOTE: can we implement Encodable/Decodable for SocketAddrV4?
+    SignOnReq(Timespec, String, Vec<u8>, PublicKey, [u8; 4], u16),
```
Comment (nixpulvis) [inline src/chat/msg.rs]: For the sake of consistency let's call these `SignOnRequest` and `SignOnResponse`.

Comment (nixpulvis) [inline src/chat/msg.rs]: I totally agree about the encode and decide for `SocketAddrV4`. Let's make that a thing.

File: src/chat/msg.rs [hunk 2]
```diff
@@ -40,6 +53,16 @@ impl Message {
         Message::Text(time::get_time(), txt.into())
     }
 
+    /// Returns a new sign on request messages.
+    pub fn sign_on_req(username: &str, hash: Vec<u8>, pk: PublicKey, ip: [u8; 4], port: u16) -> Message {
```
Comment (nixpulvis) [inline src/chat/msg.rs]: Regardless of the underlying implementation at the moment, this should take a `SocketAddr`. I'm ok with this panicking on v6 addresses honestly. Or if it's easy to serialize the `SocketAddr` type then that is obviously the best call.

File: src/chat/msg.rs [hunk 3]
```diff
@@ -48,7 +71,16 @@ impl Message {
             return Err(Error::InvalidBytes)
         }
 
+        // TODO MUST HANDLE ERRORS HERE, DON"T WANT RANDOM MESSAGE CRASHING IT
         match buf[0] {
```
Comment (nixpulvis) [inline src/chat/msg.rs]: Can't we replace the parse function with decode? And implement packets from message in terms of encode?

File: src/chat/msg.rs [hunk 4]
```diff
@@ -137,6 +189,36 @@ mod test {
         }
     }
 
+    // TODO Implment eq instead half effort matching equality tests, fully check equivalence
```
Comment (nixpulvis) [inline src/chat/msg.rs]: I'm actually totally ok with messages not implementing equality.

File: src/crypt/aes.rs [hunk 5]
```diff
@@ -29,7 +30,7 @@ impl Key {
     /// let key = Key::new(&[3; 16], &[1; 16]);
     /// ```
     // TODO: Should return a Result.
-    pub fn new(key: &[u8], iv: &[u8]) -> Result<Key> {
+    pub fn new(key: &[u8], iv: &[u8]) -> Result<Key, Error> {
```
Comment (nixpulvis) [inline src/crypt/aes.rs]: This should not return a different type of result, why was this change needed?

File: src/crypt/cert.rs [hunk 6]
```diff
@@ -34,6 +34,10 @@ impl TBSCertificate {
         }
     }
 
+    pub fn subject(&self) -> String {
```
Comment (nixpulvis) [inline src/crypt/cert.rs]: Do we really need to clone here? Why not return a `&str`.

File: src/crypt/cert.rs [hunk 7]
```diff
@@ -63,6 +67,11 @@ impl Certificate {
         }
     }
 
+    // Clone??
```
Comment (nixpulvis) [inline src/crypt/cert.rs]: See above.

File: src/chat/msg.rs (hunk 3 above)
Comment (zachferland) [inline src/chat/msg.rs]: yeah that is what I was thinking, I'll go ahead doing that

File: src/chat/msg.rs (hunk 2 above)
Comment (zachferland) [inline src/chat/msg.rs]: yeah good point. Think I may leave the 'SignOnRequest' struct as is for the time being though, since it takes no additional work to serialize it in that form. Not sure why 'SocketAddr' type doesn't implement the traits 'Encodable' and 'Decodable' already

File: src/crypt/cert.rs (hunk 6 above)
Comment (zachferland) [inline src/crypt/cert.rs]: I shouldn't, yup should be ptr

File: src/crypt/aes.rs (hunk 5 above)
Comment (zachferland) [inline src/crypt/aes.rs]: Took out our your shorted for Result to implement `encode` and 'decode'. Was that not necessary?

File: src/chat/msg.rs (hunk 2 above)
Comment (nixpulvis) [inline src/chat/msg.rs]: except we still don't have the code to go from bytes to a SocketAddr.

File: src/crypt/aes.rs (hunk 5 above)
Comment (nixpulvis) [inline src/crypt/aes.rs]: Ah, I liked the type alias. But that's ok, but why would encode and decode need this not be defined.

File: src/chat/msg.rs (hunk 2 above)
Comment (zachferland) [inline src/chat/msg.rs]: hmmm yeah, true. So that is a no go, will just have to implement encode/decode then

File: src/crypt/aes.rs (hunk 5 above)
Comment (zachferland) [inline src/crypt/aes.rs]: not really sure, I didn't think it would be issue, but then I had some errors I could not really follow. Didn't give it much thought and continued.

**Comment by nixpulvis** [comment]
Rebased, and merges fixed. I deleted a few tests in the process.
